MekHQ/src/mekhq/gui/adapter/TOEMouseAdapter.java

```diff
@@ -441,37 +443,352 @@ private void maybeShowPopup(MouseEvent e) {
                         menuItem.addActionListener(this);
                         menuItem.setEnabled(true);
                         popup.add(menuItem);
                     }
                     menu = new JMenu("Add Unit");
                     menu.setEnabled(false);
-                    // only add units that have commanders
+                    HashMap<String, JMenu> unitTypeMenus = new HashMap<String, JMenu>();
+                    HashMap<String, JMenu> weightClassForUnitType = new HashMap<String, JMenu>();
+                    
+                    // TODO: Doesn't currently account for Support Vees of any type
+                    for (int i = 0; i < UnitType.SIZE; i++)
+                    {
+                        String unittype = UnitType.getTypeName(i);
+                        String displayname = UnitType.getTypeDisplayableName(i);
+                        unitTypeMenus.put(unittype, new JMenu(displayname));
+                        unitTypeMenus.get(unittype).setName(unittype);
+                        unitTypeMenus.get(unittype).setEnabled(false);
+                        for (int j = 0; j < EntityWeightClass.getWeightLimitByType(unittype).length; j++) {
+                            double tonnage = EntityWeightClass.getWeightLimitByType(unittype)[j];
+                            // Skip over the padding 0s
+                            if (tonnage == 0) {
+                                continue;
+                            }
+                            
+                            int weightClass = EntityWeightClass.getWeightClass(tonnage, unittype);
+                            String displayname2 = EntityWeightClass.getClassName(weightClass, unittype, false);
+                            String weightClassMenuName = unittype + "_"
+                                    + EntityWeightClass.getClassName(weightClass, unittype, false);
+                            weightClassForUnitType.put(weightClassMenuName, new JMenu(displayname2));
+                            weightClassForUnitType.get(weightClassMenuName).setName(weightClassMenuName);
+                            weightClassForUnitType.get(weightClassMenuName).setEnabled(false);
+                        }
+                    }
+                    
+                    // Only add units that have commanders
                     // Or Gun Emplacements!
-                    for (Unit u : gui.getCampaign().getUnits()) {
+                    // TODO: Or Robotic Systems!
+                    JMenu unsorted = new JMenu("Unsorted");
+                    for (Unit u : gui.getCampaign().getUnits(true, true)) {
+                        String type = UnitType.determineUnitType(u.getEntity());
+                        String className = u.getEntity().getWeightClassName();
                         if (null != u.getCommander()) {
                             Person p = u.getCommander();
```

Comment:
Interior contents of this if block and the gunEmplacement block below are nearly identical except for menu item display name. Let's condense.

Reply:
That's premature optimization I don't want to do until post-stable. For now this works and I don't want to risk introduction of new bugs so close to the stable. Post-stable I plan on making sub-methods to handle the majority of this work.